Have Cypress run axe checks in WCAG 2 rulesets

bkjohnson opened the pull request:
## Description

### Addresses https://github.com/department-of-veterans-affairs/va.gov-team/issues/13647

This adds configuration to the Cypress axe checker so that it will throw errors for `serious`, `minor`, and `trivial` a11y violations, which are included in the WCAG 2 Level A and WCAG 2 Level AA rulesets. Introducing these new rules causes some existing app code to fail the stricter axe checks in Cypress, so this PR also adds a _temporary_ exception which will allow those apps to use the older less-strict checks until they fix the code that causes the violations. Once all apps have fixed the accessibility violations, the code that allows for the exception will be removed.

## Testing done

Cypress tests


## Screenshots

## Acceptance criteria
- [x] Cypress form tests run normally
- [x] Tests break on any violation impact (critical, serious, minor, trivial)
- [ ] Tests do NOT break on best-practice rules
- [ ] Tests do NOT break on experimental rules

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
Have Cypress run axe checks in WCAG 2 rulesets

U-DON commented on src/platform/testing/e2e/cypress/support/commands/axeCheck.js:
```diff
@@ -33,6 +33,10 @@ Cypress.Commands.add('axeCheck', (context = 'main') => {
 
   const options = {
     includedImpacts: ['critical'],
```
Do we still need this setting? I would've expected this line to prevent errors on anything below critical severity, but from your example, maybe it doesn't affect anything. In which case, it seems like removing it would prevent confusion down the road.

bkjohnson replied:
@1Copenut do you have any input here?

1Copenut replied:
The `includedImpacts` key and array need to be removed. This limits what level of violations we're looking for, which is not what we want. The added lines below frame the proper bounds checking.

bkjohnson committed:
Remove includedImpacts array

bkjohnson committed:
Try to disable axeChecking for some Cypress tests

bkjohnson committed:
Try to prevent Cypress axe failures in all-claims

Mottie commented on src/applications/disability-benefits/all-claims/tests/all-claims.cypress.spec.js:
```diff
@@ -14,12 +14,12 @@ const testConfig = createTestConfig(
-      'newOnly-test',
-      'secondary-new-test.json',
+      // 'newOnly-test',
+      // 'secondary-new-test.json',
```
What axe errors are showing up that require disabling these tests?

bkjohnson replied:
You can view the complete log to get this:

http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/13647-cypress-axe-fix/5/pipeline/

```
[2020-09-21T20:32:06.027Z]   Running:  [90mapplications/disability-benefits/all-claims/tests/all-claims.cypress[39m           [90m(5 of 48)[39m
[2020-09-21T20:32:06.027Z]             [90m.spec.js[39m                                                                                
[2020-09-21T20:32:06.585Z] [info] [webpackbar] Compiling Webpack
[2020-09-21T20:32:11.816Z] [success] [webpackbar] Webpack: Compiled successfully in 4.83s
[2020-09-21T20:32:15.967Z] 
[2020-09-21T20:32:15.967Z] [0m[0m
[2020-09-21T20:32:15.967Z] [0m  21-526EZ disability compensation claim form[0m
[2020-09-21T20:32:15.967Z] [0m    full-781-781a-8940-test.json[0m
[2020-09-21T20:33:23.554Z]     [32m  ✓[0m[90m fills the form[0m[31m (67092ms)[0m
[2020-09-21T20:33:23.555Z] [0m    maximal-test[0m
[2020-09-21T20:33:35.702Z] 1 accessibility violation was detected
[2020-09-21T20:33:35.702Z] ┌─────────┬──────────────────┬───────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────┐
[2020-09-21T20:33:35.702Z] │ (index) │        id        │  impact   │                                                description                                                │ nodes │
[2020-09-21T20:33:35.702Z] ├─────────┼──────────────────┼───────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────┤
[2020-09-21T20:33:35.702Z] │    0    │ 'color-contrast' │ 'serious' │ 'Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds' │   2   │
[2020-09-21T20:33:35.702Z] └─────────┴──────────────────┴───────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────┘
[2020-09-21T20:33:37.581Z]     [31m  1) fills the form[0m
[2020-09-21T20:33:37.581Z] [0m    maximal-bdd-test[0m
[2020-09-21T20:33:37.581Z]     [36m  - fills the form[0m
[2020-09-21T20:33:37.581Z] [0m    minimal-test[0m
[2020-09-21T20:33:52.398Z]     [32m  ✓[0m[90m fills the form[0m[31m (13188ms)[0m
[2020-09-21T20:33:52.398Z] [0m    minimal-bdd-test[0m
[2020-09-21T20:33:52.398Z]     [36m  - fills the form[0m
[2020-09-21T20:33:52.398Z] [0m    newOnly-test[0m
[2020-09-21T20:34:02.322Z] 1 accessibility violation was detected
[2020-09-21T20:34:02.322Z] ┌─────────┬──────────────────┬───────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────┐
[2020-09-21T20:34:02.322Z] │ (index) │        id        │  impact   │                                                description                                                │ nodes │
[2020-09-21T20:34:02.322Z] ├─────────┼──────────────────┼───────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────┤
[2020-09-21T20:34:02.322Z] │    0    │ 'color-contrast' │ 'serious' │ 'Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds' │   3   │
[2020-09-21T20:34:02.322Z] └─────────┴──────────────────┴───────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────┘
[2020-09-21T20:34:04.826Z]     [31m  2) fills the form[0m
[2020-09-21T20:34:04.826Z] [0m    secondary-new-test.json[0m
[2020-09-21T20:34:16.973Z] 1 accessibility violation was detected
[2020-09-21T20:34:16.973Z] ┌─────────┬──────────────────┬───────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────┐
[2020-09-21T20:34:16.973Z] │ (index) │        id        │  impact   │                                                description                                                │ nodes │
[2020-09-21T20:34:16.973Z] ├─────────┼──────────────────┼───────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────┤
[2020-09-21T20:34:16.973Z] │    0    │ 'color-contrast' │ 'serious' │ 'Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds' │   1   │
[2020-09-21T20:34:16.973Z] └─────────┴──────────────────┴───────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────┘
```

bkjohnson replied:
Disabling those tests doesn't seem incredibly reliable though, so I'm going to look into modifying the cypress helpers that the platform provides so that they can opt out of the axe checks.

Mottie replied:
Ahh, yes. I had to disable those contrast errors when I updated axe for Nightwatch

1Copenut replied:
It looks like two rules were throwing failures:
* `color-contrast` which is probably a false positive. Those are tough to get right in headless environments and I'd be okay suppressing this rule.
* `duplicate-id` is more problematic. Some ARIA pairings rely on unique IDs to map themselves properly. There was an issue ticket for this, do you have any recollection @Mottie ?

Mottie replied:
It _should_ have been fixed, but I added a comment that I was still seeing it - https://github.com/department-of-veterans-affairs/va.gov-team/issues/12674

bkjohnson committed:
Revert "Try to prevent Cypress axe failures in all-claims"

This reverts commit 5b231eca14ed15de1201c5b85e583d488003af6d.

bkjohnson committed:
Allow form tester to skip axeCheck

bkjohnson committed:
Revert "Allow form tester to skip axeCheck"

This reverts commit deb41a905146eefaf4e1dc6b38fccfb10614c56a.

U-DON commented on src/platform/testing/e2e/cypress/support/form-tester/index.js:
```diff
@@ -490,6 +491,7 @@ const testForm = testConfig => {
     setup = () => {},
     setupPerTest = () => {},
     skip,
+    runAxeCheck = true,
```
In the `beforeEach`, I would `cy.wrap(runAxeCheck).as('runAxeCheck')` and then as needed, do this:

```
cy.get('@runAxeCheck')
  .then(runAxeCheck => if (runAxeCheck) cy.axeCheck());
```

bkjohnson committed:
Try to allow for very granular axeCheck exceptions

The goal here is to allow the few apps that break under the stricter
rulesets to be "exceptions" for now, until the app code is fixed and
they can run the stricter ruleset

bkjohnson committed:
Specify exception instead of disabling axeCheck

U-DON commented on src/applications/facility-locator/tests/e2e/gaEvents.cypress.spec.js:
```diff
@@ -16,8 +16,8 @@ describe('Google Analytics FL Events', () => {
-      cy.injectAxe();
-      cy.axeCheck();
+      // cy.injectAxe();
+      // cy.axeCheck();
```
I don't know which is more desirable, but you may [pass a `skkipFailures` option](https://github.com/avanslaars/cypress-axe#skipfailures-optional-defaults-to-false) to continue running aXe checks without interrupting the build but still output failures in the console.

bkjohnson committed:
Try to fix error involving undefined arg

bkjohnson committed:
Add a default param

bkjohnson committed:
Allow Cypress form tester to pass the exception

bkjohnson committed:
Use the exception for fl-mobile tests

bkjohnson committed:
Add exception for hca cypress tests

bkjohnson committed:
Add exception for disability-benefits/all-claims

bkjohnson committed:
Try to add exception for facility-locator search

bkjohnson commented on src/platform/testing/e2e/cypress/support/commands/axeCheck.js:
```diff
@@ -28,12 +28,20 @@ const processAxeCheckResults = violations => {
+Cypress.Commands.add('axeCheck', (context = 'main', tempOptions = {}) => {
+  const { _13647Exception } = tempOptions;
   Cypress.log();
 
-  const options = {
-    includedImpacts: ['critical'],
-  };
+  const options = _13647Exception
+    ? {
+        includedImpacts: ['critical'],
+      }
+    : {
+        runOnly: {
+          type: 'tag',
+          values: ['wcag2a', 'wcag2aa'],
+        },
+      };
```
Running axe checks using the `wcag2a` and `wcag2aa` rulesets cause some existing cypress checks to break for a few apps. This is set up so that we can have the stricter checks in place immediately, but still allow the previous level of axe checking on the few apps which fail under the stricter checks.

bkjohnson commented on src/platform/testing/e2e/cypress/support/form-tester/index.js:
```diff
@@ -135,8 +135,8 @@ const addNewArrayItem = $form => {
-const performPageActions = pathname => {
-  cy.axeCheck();
+const performPageActions = (pathname, _13647Exception = false) => {
+  cy.axeCheck('main', { _13647Exception });
```
These changes to the form-tester can be removed once the apps can pass the stricter axe checks.

mpelzsherman commented on src/applications/facility-locator/tests/e2e/mobile.cypress.spec.js:
```diff
@@ -1,7 +1,7 @@
-Cypress.Commands.add('checkSearch', () => {
-  cy.axeCheck();
+Cypress.Commands.add('checkSearch', ({ _13647Exception = false }) => {
+  cy.axeCheck('main', { _13647Exception });
```
What does "13647Exception" mean?

bkjohnson replied:
It means that the app is being given an exception for the stricter axe checks as part of ticket https://github.com/department-of-veterans-affairs/va.gov-team/issues/13647.

This is only temporary, and once this PR is merged I will set up separate tickets/PRs for each individual app teams to fix whatever is causing the axe checks to fail.

cvalarida replied:
To be clear, the name is intentionally wonky; it's not meant to be used by application teams writing new tests.

mpelzsherman replied:
I see. Well even so, maybe `temporaryException` would be more self-documenting?

mpelzsherman replied:
Will it be up to the facilities team to make the tests pass without the exception?

bkjohnson replied:
> Will it be up to the facilities team to make the tests pass without the exception?

Yes. Once I merge this I will open a ticket, and even a PR for a branch where I remove the exception so that the failures are visible.

cvalarida reviewed (approved):
The toggle to use the old vs new options looks good. Once you make those tickets / PRs, be sure to link to this ticket so we can follow the audit trail. :+1:

1Copenut reviewed (approved):
LGTM. Thanks @bkjohnson for the fix and strategy to bring testing back in as errors get fixed.

mpelzsherman reviewed (approved):
In the future, it would be great if we could get a little heads up about this sort of thing before being asked to review a PR.
